src/string/grapheme.cr

```diff
@@ -0,0 +1,64 @@
+require "./grapheme/grapheme"
+
+class String
+  # Returns this string split into Unicode extended grapheme clusters.
+  #
+  # `Grapheme` clusters correspond to "user-perceived characters" and are defined
+  # in [Unicode Standard Annex #29](https://unicode.org/reports/tr29/). A cluster
+  # can consist of multiple code points which together form a single glyph.
+  #
+  # ```
+  # "a👍🏼à".graphemes # => [String::Grapheme('a'), String::Grapheme("👍🏼"), String::Grapheme("à")]
+  # ```
+  #
+  # * `#each_grapheme` iterates the grapheme clusters without allocating an array
+  def graphemes : Array(Grapheme)
+    graphemes = [] of Grapheme
+    each_grapheme do |grapheme|
+      graphemes << grapheme
+    end
+    graphemes
+  end
+
+  # Yields each Unicode extended grapheme cluster in this string.
+  #
+  # `Grapheme` clusters correspond to "user-perceived characters" and are defined
+  # in [Unicode Standard Annex #29](https://unicode.org/reports/tr29/). A cluster
+  # can consist of multiple code points which together form a single glyph.
+  #
+  # ```
+  # "a👍🏼à".each_grapheme do |cluster|
+  #   p! cluster
+  # end
+  # ```
+  #
+  # * `#graphemes` collects all grapheme clusters in an array
+  def each_grapheme(& : Grapheme -> _) : Nil
+    each_grapheme_boundary do |range, last_char|
+      yield Grapheme.new(self, range, last_char)
+    end
+  end
+
+  def grapheme_size : Int32
```

Comment:
Is this intended to be public API?

Reply:
Yes. I'll need to add docs for this.

Reply:
Sounds good!

But... is there a use-case for this particular method?

Reply:
I think it can be useful as an approximation of string width. It's not accurate because the display width of each graphemes is unknown. But it's likely a better indicator than `String#size`.

Reply:
Oh, that makes total sense 👍

Reply:
For getting unicode String width, please consider taking a look at my another shard [uni_char_width](https://github.com/naqvis/uni_char_width) which calculates the width of unicode char and String per Unicode specs.

Reply:
(It still misses the doc)